File: src/dexter/tools/finance/news.py
from langchain.tools import tool
from typing import Optional
from pydantic import BaseModel, Field
from dexter.tools.finance.api import call_api

class NewsInput(BaseModel):
    """Input for get_news."""
    ticker: str = Field(..., description="The stock ticker symbol to fetch news for. For example, 'AAPL' for Apple.")
    start_date: Optional[str] = Field(default=None, description="The start date to fetch news from (YYYY-MM-DD).")
    end_date: Optional[str] = Field(default=None, description="The end date to fetch news to (YYYY-MM-DD).")
    limit: int = Field(default=10, description="The number of news articles to retrieve. Max is 100.")

@tool(args_schema=NewsInput)
def get_news(
    ticker: str,
    start_date: Optional[str] = None,
    end_date: Optional[str] = None,
    limit: int = 5,
) -> dict:
    """
    Retrieves recent news articles for a given company ticker,
    covering financial announcements, market trends, and other significant events.
    Useful for staying up-to-date with market-moving information and investor sentiment.
    """
    params = {
        "ticker": ticker,
        "limit": limit,
    }
    if start_date:
        params["start_date"] = start_date
    if end_date:
        params["end_date"] = end_date
    
    data = call_api("/news/", params)
    return data.get("news", [])

File: src/tools/finance/news.ts
import { DynamicStructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import { api } from './api.js';
import { formatToolResult } from '../types.js';
import { validateLimit, validateTicker } from './validation.js';
import { TTL_15M } from './utils.js';

const CompanyNewsInputSchema = z.object({
  ticker: z
    .string()
    .describe("The stock ticker symbol to fetch company news for. For example, 'AAPL' for Apple."),
  limit: z
    .number()
    .default(5)
    .describe('Maximum number of news articles to return (default: 5, max: 10).'),
});

export const getCompanyNews = new DynamicStructuredTool({
  name: 'get_company_news',
  description:
    'Retrieves recent company news headlines for a stock ticker, including title, source, publication date, and URL. Use for company catalysts, price move explanations, press releases, and recent announcements.',
  schema: CompanyNewsInputSchema,
  func: async (input) => {
    const ticker = validateTicker(input.ticker);
    const limit = validateLimit(input.limit, { fieldName: 'limit', min: 1, max: 10 });

    const params: Record<string, string | number | undefined> = {
      ticker,
      limit,
    };
    const { data, url } = await api.get('/news', params, { cacheable: true, ttlMs: TTL_15M });
    return formatToolResult((data.news as unknown[]) || [], [url]);
  },
});
